singer/src/main/java/com/pinterest/singer/writer/s3/S3Writer.java

```diff
@@ -0,0 +1,438 @@
+     * */
+    public String getS3FileFormat() {
+        // FORMAT: <log_name>/<service_fleet>/<host>/<log_dir>/<custom_filename>.<timestamp>
+        String logName = logStream.getSingerLog().getSingerLogConfig().getName();
+        List<String> hostPrefixes = getHostnamePrefixes("-");
+        String serviceFleet = hostPrefixes.get(hostPrefixes.size() - 2);
+        String host = extractHostSuffix(HOSTNAME);
+        String logDir = logStream.getFullPathPrefix().substring(1);
+        String customFilename = s3WriterConfig.getFileNameFormat();
+        String timestamp = formatter.format(new Date());
+        LOG.info("Uploading the file: " + logName + "/" + serviceFleet + "/" + host + "/" + logDir + "/" + customFilename +
+                "." + timestamp + " to the bucket " + bucketName + " with key prefix " + keyPrefix);
+        return logName + "/" + serviceFleet + "/" + host + "/" + logDir + "/" + customFilename + "." + timestamp;
+    }
+
+    /**
+     * Helper function to create a new file to upload to S3.
+     * */
+    private File createNewS3File() {
+        String newFileName = s3WriterConfig.getFileNameFormat() + "." + formatter.format(new Date());
+        return new File(BUFFER_DIR, newFileName);
+    }
+
+    /**
+     * Ends the commit process by flushing the buffer and handling the buffer file if it exceeds the maximum file size.
+     *
+     * @param numLogMessagesRead the number of log messages read
+     * @param isDraining whether the system is in a draining state
+     * @throws LogStreamWriterException if an error occurs while ending the commit
+     */
+    public synchronized void endCommit(int numLogMessagesRead, boolean isDraining) throws LogStreamWriterException {
+        try {
+            synchronized (objLock) {
+                if (bufferFile.length() >= maxFileSizeMB * mB) {
+                    if (uploadFuture != null) {
+                        uploadFuture.cancel(true);
+                    }
+                    File newFile = createNewS3File();
+                    String fileFormat = getS3FileFormat();
+                    bufferedOutputStream.close();
+                    uploadDiskBufferedFileToS3(newFile, fileFormat);
+                    scheduleUploadTask();
+                }
+            }
+        } catch (IOException e) {
+            throw new LogStreamWriterException("Failed to end commit", e);
+        }
+    }
+
+    /**
+     * This method should not be used as it is Deprecated. Use comittable write method instead.
+     * @param messages The LogMessages to be written.
+     * @throws LogStreamWriterException
+     */
+    @Deprecated
+    public void writeLogMessages(List<LogMessage> messages) throws LogStreamWriterException {
+        throw new LogStreamWriterException("writeLogMessages is not supported. Use writeLogMessagesToCommit instead.");
+    }
+
+
+    /**
+     * This method is called per LogStreamConfig, not for all of Singer.
```

Comment:
nit: don't think this comment is necessary. Perhaps it would be better to explain what is being cleaned up?

Reply:
Got it, will do